File: modules/nitro-p2p/ios/HybridP2p.swift
import Foundation
import Network
import NitroModules

// ─────────────────────────────────────────────────────────────────────────────
// HybridP2P.swift
//
// Wire protocol (identical to Kotlin counterpart):
//   Discovery  → UDP multicast  239.255.42.1:45679  JSON beacon every 1.5s
//   Transport  → TCP            [UInt32 BE length][UTF-8 payload]
//   Handshake  → first TCP msg  "__HS__{"id":…,"name":…,"port":…}"
// ─────────────────────────────────────────────────────────────────────────────

private let kBeaconHost      = "239.255.42.1"
private let kBeaconPort: UInt16 = 45679
private let kBeaconInterval: TimeInterval = 1.5
private let kBeaconTTL:      TimeInterval = 6.0

class HybridP2P: HybridP2PSpec {
    // ─── Identity ─────────────────────────────────────────────────────────────
    private var deviceId   = UUID().uuidString
    private var deviceName = UIDevice.current.name
    private var tcpPort: UInt16 = 45678
    
    // ─── Subscriber registry ──────────────────────────────────────────────────
    // Key: subscription ID (returned to caller)
    // Value: the callback closure
    // Protected by `q` — all reads/writes happen on that queue.
    private var subscribers: [Double: (P2PEvent) -> Void] = [:]
    private var nextSubscriberId: Double = 0   // auto-increment, never reused
    
    // ─── Network state ────────────────────────────────────────────────────────
    private var tcpListener:   NWListener?
    private var udpListener:   NWListener?
    private var udpSendConn:   NWConnection?
    private var beaconTimer:   Timer?
    private var expiryTimer:   Timer?
    private var discoveredMap: [String: DiscoveredPeer] = [:]
    private var connections:   [String: ManagedConnection] = [:]
    
    // ─── Dispatch queue ───────────────────────────────────────────────────────
    // Single serial queue: all network I/O, subscriber registry, and state
    // mutations happen here. Callbacks are dispatched to main before firing.
    private let q = DispatchQueue(label: "com.poolgame.p2p", qos: .userInteractive)
    
    
    func setDeviceName(name: String) throws {
        deviceName = name
    }
    
    func start(servicePort: Double?) throws {
        tcpPort = servicePort.map { UInt16($0) } ?? 45678
        q.async { [weak self] in
          self?.startTCPListener()
          self?.startUDPListener()
          self?.startBeaconing()
          self?.startExpiryTimer()
        }
    }
    
    func stop() throws {
        q.async { [weak self] in
          guard let self else { return }
          self.beaconTimer?.invalidate()
          self.expiryTimer?.invalidate()
          self.udpListener?.cancel()
          self.udpSendConn?.cancel()
          self.tcpListener?.cancel()
          self.connections.values.forEach { $0.connection.cancel() }
          self.connections.removeAll()
          self.discoveredMap.removeAll()
          // Subscribers are NOT cleared on stop — they survive a restart.
        }
    }
    
    func connect(peerId: String) throws {
        q.async { [weak self] in
          guard let self, self.connections[peerId] == nil else { return }
          guard let peer = self.discoveredMap[peerId] else {
              self.emit(
                P2PEvent.sixth(.init(type: .error, code: "PEER_NOT_FOUND", message: "PEER_NOT_FOUND")))
            return
          }
          let endpoint = NWEndpoint.hostPort(
            host: NWEndpoint.Host(peer.host),
            port: NWEndpoint.Port(rawValue: UInt16(peer.port))!
          )
          let conn = NWConnection(to: endpoint, using: .tcp)
          self.registerConnection(conn, peerId: peerId, knownPeer: peer.asP2PPeer())
          conn.start(queue: self.q)
        }
    }
    
    func disconnect(peerId: String) throws {
        q.async { [weak self] in
          self?.connections[peerId]?.connection.cancel()
          self?.connections.removeValue(forKey: peerId)
        }
    }
    
    func sendToPeer(peerId: String, data: String) throws {
        q.async { [weak self] in self?.connections[peerId]?.send(data) }
    }
    
    func broadcast(data: String) throws {
        q.async { [weak self] in self?.connections.values.forEach { $0.send(data) } }
    }
    
    func getDiscoveredPeers() throws -> [P2PPeer] {
        discoveredMap.values.map { $0.asP2PPeer() }
    }
    
    func getConnectedPeers() throws -> [P2PPeer] {
        connections.values.compactMap { $0.peer }
    }
    
    func subscribe(callback: @escaping (P2PEvent) -> Void) throws -> Double {
        // Mutations to `subscribers` happen on `q` to keep them thread-safe,
        // but we need to return the ID synchronously to the JS caller.
        // `sync` is safe here because subscribe() is never called FROM `q`.
        return q.sync {
          let id = nextSubscriberId
          nextSubscriberId += 1
          subscribers[id] = callback
          return id
        }
    }
    
    func unsubscribe(id: Double) throws {
        q.async { [weak self] in
          self?.subscribers.removeValue(forKey: id)
        }
    }

    // Every network event calls `emit(_:)`. It iterates the subscriber map
    // and dispatches each callback onto the main thread so callers never have
    // to think about threading.

    private func emit(_ event: P2PEvent) {
      // Snapshot the subscriber values so we can release the queue lock quickly.
      let callbacks = subscribers.values
      DispatchQueue.main.async {
        callbacks.forEach { $0(event) }
      }
    }
    
    
    private func startTCPListener() {
      guard let listener = try? NWListener(using: .tcp, on: NWEndpoint.Port(rawValue: tcpPort)!) else {
          self.emit(.sixth(.init(type: .error, code: "TCP_BIND_FAILED", message: "Cannot bind port \(tcpPort)")))
        return
      }
      tcpListener = listener
      listener.newConnectionHandler = { [weak self] conn in
        // Inbound connection: peerId unknown until handshake arrives
        self?.registerConnection(conn, peerId: nil, knownPeer: nil)
        conn.start(queue: self!.q)
      }
      listener.start(queue: q)
    }

    // MARK: - UDP multicast listener

    private func startUDPListener() {
      let params = NWParameters.udp
      params.allowLocalEndpointReuse = true
      guard let listener = try? NWListener(using: params, on: NWEndpoint.Port(rawValue: kBeaconPort)!) else {
        
          self.emit(.sixth(.init(type: .error, code: "UDP_BIND_FAILED", message: "Cannot bind UDP \(kBeaconPort)")))
        return
      }
      udpListener = listener
      listener.newConnectionHandler = { [weak self] conn in
        conn.start(queue: self!.q)
        self?.receiveUDP(on: conn)
      }
      listener.start(queue: q)
    }

    private func receiveUDP(on conn: NWConnection) {
      conn.receiveMessage { [weak self] data, _, _, error in
        if let data, let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
          self?.handleBeacon(json)
        }
        if error == nil { self?.receiveUDP(on: conn) }
      }
    }

    // MARK: - Beaconing

    private func startBeaconing() {
      let endpoint = NWEndpoint.hostPort(
        host: NWEndpoint.Host(kBeaconHost),
        port: NWEndpoint.Port(rawValue: kBeaconPort)!
      )
      udpSendConn = NWConnection(to: endpoint, using: .udp)
      udpSendConn?.start(queue: q)
      sendBeacon()
      beaconTimer = Timer.scheduledTimer(withTimeInterval: kBeaconInterval, repeats: true) { [weak self] _ in
        self?.q.async { self?.sendBeacon() }
      }
      RunLoop.main.add(beaconTimer!, forMode: .common)
    }

    private func sendBeacon() {
      let payload: [String: Any] = ["id": deviceId, "name": deviceName, "port": tcpPort]
      guard let data = try? JSONSerialization.data(withJSONObject: payload) else { return }
      udpSendConn?.send(content: data, completion: .idempotent)
    }

    private func handleBeacon(_ json: [String: Any]) {
      guard
        let peerId   = json["id"]   as? String,
        let peerName = json["name"] as? String,
        let peerPort = json["port"] as? Int,
        peerId != deviceId
      else { return }

      let isNew = discoveredMap[peerId] == nil
      discoveredMap[peerId] = DiscoveredPeer(id: peerId, name: peerName, host: "", port: peerPort, lastSeen: Date())
        if isNew {
            emit(.first(.init(type: .peerdiscovered, peer: discoveredMap[peerId]!.asP2PPeer())))
        }
    }

    // MARK: - Peer expiry

    private func startExpiryTimer() {
      expiryTimer = Timer.scheduledTimer(withTimeInterval: 2.0, repeats: true) { [weak self] _ in
        self?.q.async {
          let now     = Date()
          let expired = self?.discoveredMap.filter { now.timeIntervalSince($0.value.lastSeen) > kBeaconTTL } ?? [:]
          expired.keys.forEach { id in
            self?.discoveredMap.removeValue(forKey: id)
              self?.emit(.second(.init(type: .peerlost, peerId: id)))
          }
        }
      }
      RunLoop.main.add(expiryTimer!, forMode: .common)
    }

    // MARK: - TCP connection lifecycle

    private func registerConnection(_ conn: NWConnection, peerId: String?, knownPeer: P2PPeer?) {
      let mc = ManagedConnection(connection: conn, peer: knownPeer)
      if let pid = peerId { connections[pid] = mc }

      conn.stateUpdateHandler = { [weak self, weak mc] state in
        guard let self, let mc else { return }
        switch state {
        case .ready:
          // Send our handshake immediately so the remote learns our identity
          let hs: [String: Any] = ["id": self.deviceId, "name": self.deviceName, "port": Int(self.tcpPort)]
          if let data = try? JSONSerialization.data(withJSONObject: hs),
             let str  = String(data: data, encoding: .utf8) {
            mc.send("__HS__" + str)
          }
            if let peer = mc.peer { emit(P2PEvent.third(.init(type: .peerconnected, peer: peer))) }
          self.receiveFramed(mc)
        case .failed(let err):
            self.emit(.sixth(.init(type: .error, code: "CONNECTION_FAILED", message: err.localizedDescription)))
          self.cleanupConnection(mc)
        case .cancelled:
          self.cleanupConnection(mc)
        default: break
        }
      }
    }

    private func receiveFramed(_ mc: ManagedConnection) {
      mc.connection.receive(minimumIncompleteLength: 4, maximumLength: 4) { [weak self, weak mc] data, _, _, error in
        guard let self, let mc, error == nil, let data, data.count == 4 else { return }
        let length = Int(data.withUnsafeBytes { $0.load(as: UInt32.self).bigEndian })
        guard length > 0, length < 1_000_000 else { return }

        mc.connection.receive(minimumIncompleteLength: length, maximumLength: length) { [weak self, weak mc] payload, _, _, error in
          guard let self, let mc, error == nil,
                let payload, let str = String(data: payload, encoding: .utf8) else { return }

          if str.hasPrefix("__HS__"),
             let json    = try? JSONSerialization.jsonObject(with: Data(str.dropFirst(6).utf8)) as? [String: Any],
             let peerId  = json["id"]   as? String,
             let peerName = json["name"] as? String,
             let peerPort = json["port"] as? Int {
            let peer     = P2PPeer(id: peerId, name: peerName, host: "", port: Double(peerPort))
            mc.peer      = peer
            self.connections[peerId] = mc
              
              emit(P2PEvent.third(.init(type: .peerconnected, peer: peer)))
          } else if let peerId = mc.peer?.id {
              emit(.fifth(.init(type: .messagereceived, message: .init(fromPeerId: peerId, data: str))))
          }
          self.receiveFramed(mc)
        }
      }
    }

    private func cleanupConnection(_ mc: ManagedConnection) {
      if let pid = mc.peer?.id {
        connections.removeValue(forKey: pid)
          emit(P2PEvent.second(.init(type: .peerlost, peerId: pid)))
      }
    }
    
}

private class ManagedConnection {
  let connection: NWConnection
  var peer: P2PPeer?
  init(connection: NWConnection, peer: P2PPeer?) { self.connection = connection; self.peer = peer }

  func send(_ string: String) {
    guard let data = string.data(using: .utf8) else { return }
    var length = UInt32(data.count).bigEndian
    let prefix = Data(bytes: &length, count: 4)
    connection.send(content: prefix + data, completion: .idempotent)
  }
}

private struct DiscoveredPeer {
  let id: String; let name: String; var host: String; let port: Int; var lastSeen: Date
  func asP2PPeer() -> P2PPeer { P2PPeer(id: id, name: name, host: host, port: Double(port)) }
}

File: modules/nitro-p2p/android/src/main/java/com/margelo/nitro/nitrop2p/HybridP2p.kt
package com.margelo.nitro.nitrop2p

import android.content.Context
import android.net.wifi.WifiManager
import android.os.Handler
import android.os.Looper
import com.margelo.nitro.core.HybridObject
import org.json.JSONObject
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException
import java.net.DatagramPacket
import java.net.InetAddress
import java.net.MulticastSocket
import java.net.ServerSocket
import java.net.Socket
import java.net.SocketException
import java.net.SocketTimeoutException
import java.nio.ByteBuffer
import java.util.Date
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

// ─────────────────────────────────────────────────────────────────────────────
// HybridP2P.kt
//
// Wire protocol (identical to Swift counterpart):
//   Discovery  → UDP multicast  239.255.42.1:45679  JSON beacon every 1.5s
//   Transport  → TCP            [UInt32 BE length][UTF-8 payload]
//   Handshake  → first TCP msg  "__HS__{"id":…,"name":…,"port":…}"
// ─────────────────────────────────────────────────────────────────────────────

private const val BEACON_HOST = "239.255.42.1"
private const val BEACON_PORT = 45679
private const val BEACON_INTERVAL_MS = 1500L
private const val BEACON_TTL_MS = 6000L
private const val EXPIRY_CHECK_INTERVAL_MS = 2000L

class HybridP2P(val context: Context) : HybridP2PSpec() {

    // ─── Identity ─────────────────────────────────────────────────────────────
    private val deviceId: String = UUID.randomUUID().toString()
    private var deviceName: String = android.os.Build.MODEL
    private var tcpPort: Int = 45678

    // ─── Subscriber registry ──────────────────────────────────────────────────
    // Key: subscription ID (returned to caller)
    // Value: the callback function
    private val subscribers: ConcurrentHashMap<Double, (P2PEvent) -> Unit> = ConcurrentHashMap()
    private val nextSubscriberId: AtomicInteger = AtomicInteger(0)

    // ─── Network state ────────────────────────────────────────────────────────
    private var tcpServerSocket: ServerSocket? = null
    private var multicastSocket: MulticastSocket? = null
    private var multicastLock: WifiManager.MulticastLock? = null
    private var beaconScheduler: ScheduledExecutorService? = null
    private var expiryScheduler: ScheduledExecutorService? = null
    private val discoveredPeers: ConcurrentHashMap<String, DiscoveredPeer> = ConcurrentHashMap()
    private val connections: ConcurrentHashMap<String, ManagedConnection> = ConcurrentHashMap()
    private val networkExecutor = Executors.newCachedThreadPool()

    // ─── Main thread handler for event emission ─────────────────────────────
    private val mainHandler = Handler(Looper.getMainLooper())

    // ─── Data classes ─────────────────────────────────────────────────────────
    private data class DiscoveredPeer(
        val id: String,
        val name: String,
        var host: String,
        val port: Int,
        var lastSeen: Date
    ) {
        fun toP2PPeer(): P2PPeer = P2PPeer(id, name, host, port.toDouble())
    }

    private class ManagedConnection(
        val socket: Socket,
        var peer: P2PPeer?,
        val output: DataOutputStream
    ) {
        fun send(data: String) {
            try {
                val payload = data.toByteArray(Charsets.UTF_8)
                val length = ByteBuffer.allocate(4).putInt(payload.size).array()
                output.write(length)
                output.write(payload)
                output.flush()
            } catch (e: IOException) {
                // Connection will be cleaned up by read thread
            }
        }
    }

    // ─── Public API ─────────────────────────────────────────────────────────────

    override fun setDeviceName(name: String) {
        deviceName = name
    }

    override fun start(servicePort: Double?) {
        servicePort?.let { tcpPort = it.toInt() }
        networkExecutor.execute {
            startTCPServer()
            startUDPMulticast()
            startBeaconing()
            startExpiryTimer()
        }
    }

    override fun stop() {
        networkExecutor.execute {
            beaconScheduler?.shutdownNow()
            expiryScheduler?.shutdownNow()
            multicastSocket?.close()
            multicastLock?.release()
            tcpServerSocket?.close()
            connections.values.forEach { closeConnection(it) }
            connections.clear()
            discoveredPeers.clear()
        }
    }

    override fun connect(peerId: String) {
        networkExecutor.execute {
            if (connections[peerId] != null) return@execute

            val peer = discoveredPeers[peerId] ?: run {
                emitError("PEER_NOT_FOUND", "Peer not found: $peerId")
                return@execute
            }

            try {
                val socket = Socket(peer.host, peer.port)
                val output = DataOutputStream(socket.getOutputStream())
                val mc = ManagedConnection(socket, null, output)
                connections[peerId] = mc
                handleConnection(mc)
            } catch (e: IOException) {
                emitError("CONNECTION_FAILED", e.message ?: "Failed to connect")
            }
        }
    }

    override fun disconnect(peerId: String) {
        networkExecutor.execute {
            connections[peerId]?.let { closeConnection(it) }
            connections.remove(peerId)
            emit(P2PEvent.create(P2PPeerDisconnected(type = P2PEventType.PEERDISCONNECTED, peerId = peerId)))
        }
    }

    override fun sendToPeer(peerId: String, data: String) {
        networkExecutor.execute {
            connections[peerId]?.send(data)
        }
    }

    override fun broadcast(data: String) {
        networkExecutor.execute {
            connections.values.forEach { it.send(data) }
        }
    }

    override fun getDiscoveredPeers(): Array<P2PPeer> {
        return discoveredPeers.values.map { it.toP2PPeer() }.toTypedArray()
    }

    override fun getConnectedPeers(): Array<P2PPeer> {
        return connections.values.mapNotNull { it.peer }.toTypedArray()
    }

    override fun subscribe(callback: (P2PEvent) -> Unit): Double {
        val id = nextSubscriberId.getAndIncrement().toDouble()
        subscribers[id] = callback
        return id
    }

    override fun unsubscribe(id: Double) {
        subscribers.remove(id)
    }

    // ─── Event Emission ─────────────────────────────────────────────────────────────

    private fun emit(event: P2PEvent) {
        // Snapshot callbacks to allow lock-free iteration
        val callbacks = subscribers.values.toList()
        mainHandler.post {
            callbacks.forEach { it(event) }
        }
    }

    private fun emitError(code: String, message: String) {
        emit(P2PEvent.create(P2PErrorEvent(type = P2PEventType.ERROR, code = code, message = message)))
    }

    // ─── TCP Server ─────────────────────────────────────────────────────────────

    private fun startTCPServer() {
        try {
            tcpServerSocket = ServerSocket(tcpPort)
            while (tcpServerSocket?.isClosed == false) {
                try {
                    val socket = tcpServerSocket?.accept() ?: break
                    val output = DataOutputStream(socket.getOutputStream())
                    val mc = ManagedConnection(socket, null, output)
                    networkExecutor.execute { handleConnection(mc) }
                } catch (e: SocketException) {
                    // Socket closed, exit loop
                    break
                }
            }
        } catch (e: IOException) {
            emitError("TCP_BIND_FAILED", "Cannot bind port $tcpPort: ${e.message}")
        }
    }

    private fun handleConnection(mc: ManagedConnection) {
        // Send handshake immediately on connect
        val handshakeJson = JSONObject().apply {
            put("id", deviceId)
            put("name", deviceName)
            put("port", tcpPort)
        }
        mc.send("__HS__$handshakeJson")

        mc.peer?.let {
            emit(P2PEvent.create(P2PPeerConnected(type = P2PEventType.PEERCONNECTED, peer = it)))
        }

        // Read loop
        val input = DataInputStream(mc.socket.getInputStream())
        try {
            while (!mc.socket.isClosed) {
                // Read 4-byte length prefix (big-endian)
                val lengthBytes = ByteArray(4)
                if (input.read(lengthBytes) != 4) break
                val length = ByteBuffer.wrap(lengthBytes).int

                if (length !in 1..1_000_000) break

                // Read payload
                val payloadBytes = ByteArray(length)
                input.readFully(payloadBytes)
                val payload = String(payloadBytes, Charsets.UTF_8)

                when {
                    payload.startsWith("__HS__") -> {
                        try {
                            val jsonStr = payload.substring(6)
                            val json = JSONObject(jsonStr)
                            val peerId = json.getString("id")
                            val peerName = json.getString("name")
                            val peerPort = json.getInt("port")

                            val peer = P2PPeer(peerId, peerName, mc.socket.inetAddress.hostAddress ?: "", peerPort.toDouble())
                            mc.peer = peer
                            connections[peerId] = mc

                            // Get host from socket
                            discoveredPeers[peerId] = discoveredPeers[peerId]?.copy(
                                host = mc.socket.inetAddress.hostAddress ?: ""
                            ) ?: DiscoveredPeer(peerId, peerName, mc.socket.inetAddress.hostAddress ?: "", peerPort, Date())

                            emit(P2PEvent.create(P2PPeerConnected(type = P2PEventType.PEERCONNECTED, peer = peer)))
                        } catch (e: Exception) {
                            // Invalid handshake
                        }
                    }
                    mc.peer != null -> {
                        emit(P2PEvent.create(P2PMessageReceived(
                            type = P2PEventType.MESSAGERECEIVED,
                            message = P2PMessage(fromPeerId = mc.peer!!.id, data = payload)
                        )))
                    }
                }
            }
        } catch (e: IOException) {
            // Connection closed or error
        } finally {
            cleanupConnection(mc)
        }
    }

    private fun closeConnection(mc: ManagedConnection) {
        try {
            mc.socket.close()
        } catch (e: IOException) {
            // Ignore
        }
    }

    private fun cleanupConnection(mc: ManagedConnection) {
        mc.peer?.id?.let { peerId ->
            connections.remove(peerId)
            emit(P2PEvent.create(P2PPeerDisconnected(type = P2PEventType.PEERDISCONNECTED, peerId = peerId)))
        }
        closeConnection(mc)
    }

    // ─── UDP Multicast ─────────────────────────────────────────────────────────────

    private fun startUDPMulticast() {
        try {
            // Acquire multicast lock
            val wifiManager = context.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
            multicastLock = wifiManager.createMulticastLock("nitro-p2p").apply {
                setReferenceCounted(true)
                acquire()
            }

            // Create multicast socket
            val socket = MulticastSocket(BEACON_PORT)
            multicastSocket = socket
            val group = InetAddress.getByName(BEACON_HOST)
            socket.joinGroup(group)
            socket.soTimeout = 1000

            // Receive beacons
            val buffer = ByteArray(1024)
            val packet = DatagramPacket(buffer, buffer.size)

            while (!socket.isClosed) {
                try {
                    socket.receive(packet)
                    val jsonStr = String(packet.data, 0, packet.length, Charsets.UTF_8)
                    val json = JSONObject(jsonStr)
                    handleBeacon(json, packet.address.hostAddress ?: "")
                } catch (e: SocketTimeoutException) {
                    // Timeout, continue loop
                } catch (e: Exception) {
                    // Error receiving, continue loop
                }
            }
        } catch (e: IOException) {
            emitError("UDP_BIND_FAILED", "Cannot bind UDP multicast: ${e.message}")
        }
    }

    private fun startBeaconing() {
        beaconScheduler = Executors.newSingleThreadScheduledExecutor()
        beaconScheduler?.scheduleWithFixedDelay({
            try {
                val json = JSONObject().apply {
                    put("id", deviceId)
                    put("name", deviceName)
                    put("port", tcpPort)
                }
                val data = json.toString().toByteArray(Charsets.UTF_8)

                val group = InetAddress.getByName(BEACON_HOST)
                val packet = DatagramPacket(data, data.size, group, BEACON_PORT)
                multicastSocket?.send(packet)
            } catch (e: Exception) {
                // Ignore send errors
            }
        }, 0, BEACON_INTERVAL_MS, TimeUnit.MILLISECONDS)
    }

    private fun handleBeacon(json: JSONObject, host: String) {
        try {
            val peerId = json.getString("id")
            val peerName = json.getString("name")
            val peerPort = json.getInt("port")

            if (peerId == deviceId) return

            val isNew = discoveredPeers[peerId] == null
            discoveredPeers[peerId] = DiscoveredPeer(peerId, peerName, host, peerPort, Date())

            if (isNew) {
                emit(P2PEvent.create(P2PPeerDiscovered(type = P2PEventType.PEERDISCOVERED, peer = discoveredPeers[peerId]!!.toP2PPeer())))
            }
        } catch (e: Exception) {
            // Invalid beacon
        }
    }

    // ─── Peer Expiry ─────────────────────────────────────────────────────────────

    private fun startExpiryTimer() {
        expiryScheduler = Executors.newSingleThreadScheduledExecutor()
        expiryScheduler?.scheduleWithFixedDelay({
            val now = Date()
            val expiredIds = discoveredPeers.filter { (_, peer) ->
                now.time - peer.lastSeen.time > BEACON_TTL_MS
            }.keys

            expiredIds.forEach { id ->
                discoveredPeers.remove(id)
                emit(P2PEvent.create(P2PPeerLost(type = P2PEventType.PEERLOST, peerId = id)))
            }
        }, EXPIRY_CHECK_INTERVAL_MS, EXPIRY_CHECK_INTERVAL_MS, TimeUnit.MILLISECONDS)
    }
}
